### PR TITLE
11106 layer controller refactor

### LINKED ISSUE #11106: Layer controller refactor (take 2)
Continue work from https://github.com/CartoDB/cartodb/pull/10934

We wanted to start with UserTable, but it was not viable as it used a lot of functionality from Layer. So in #10934 we did the Layer model, which should help make the UserTable easier.

This issue will focus on the Layer controller in order to finish adding functionality to the Layer model which may be missing from the layer model unit tests.

### PR DESCRIPTION
**Pull request by javitonino**
Closes #11106

### CODE REVIEW

**Comment by gfiorav** [comment]
Test have failed! Take a look!

**Comment by gfiorav** [comment]
Fixe the obvious issue, but it seems that there are some more profound issues with tests. Take a look please!

**Comment by javitonino** [comment]
Fixed the tests, there were three instances of map without visualization in the specs. Just adding a dummy visualization works.

File: spec/requests/api/json/layers_controller_spec.rb [hunk 1]
```diff
@@ -1,276 +0,0 @@
-# encoding: utf-8
```
Comment (juanignaciosl) [inline spec/requests/api/json/layers_controller_spec.rb]: Shouldn't these tests be moved to `carto/api/layers_controller_spec`?

File: app/controllers/carto/api/layers_controller.rb [hunk 2]
```diff
@@ -4,81 +4,284 @@
...
+          ).visualization
+          unless table_visualization.has_read_permission?(current_user)
+            raise UnauthorizedError.new('You do not have permission in the layer you are trying to add')
+          end
+        end
       end
 
-      def custom_layers_by_user
-        raise RecordNotFound if @owner_user.nil?
+      def index(parent, owner: nil)
```
Comment (juanignaciosl) [inline app/controllers/carto/api/layers_controller.rb]: As `parent` is only used for `parent.layers`, `layers` is a simpler parameter candidate (and then we don't have to name this `layered_parent` or something similar 😅 ).

File: app/controllers/carto/api/layers_controller.rb [hunk 3]
```diff
@@ -4,81 +4,284 @@
...
+        destroy
+      end
+
+      def user_destroy
+        destroy
+      end
+
+      private
+
+      def validate_for_map(layer)
```
Comment (juanignaciosl) [inline app/controllers/carto/api/layers_controller.rb]: Can this be moved to model validation?

File: app/controllers/carto/api/layers_controller.rb [hunk 4]
```diff
@@ -4,81 +4,284 @@
...
+            message: 'Error creating layer',
+            errors: layer.errors.full_messages
+          )
+          raise UnprocesableEntityError.new(layer.errors.full_messages)
         end
+      end
 
-        render_jsonp layers: layers, total_entries: layers.size
+      def user_create
+        layer = Carto::Layer.new(params.slice(:kind, :options, :infowindow, :tooltip, :order))
```
Comment (juanignaciosl) [inline app/controllers/carto/api/layers_controller.rb]: Duplicated. This can be extracted to a `layer_from_param` method.

File: app/controllers/carto/api/layers_controller.rb [hunk 5]
```diff
@@ -4,81 +4,284 @@
...
+        else
+          CartoDB::Logger.info(
+            message: 'Error creating layer',
+            errors: layer.errors.full_messages
+          )
+          raise UnprocesableEntityError.new(layer.errors.full_messages)
+        end
+      end
+
+      def map_update
```
Comment (juanignaciosl) [inline app/controllers/carto/api/layers_controller.rb]: Can't these 4 methods be merged into 2?

File: app/controllers/carto/api/layers_controller.rb [hunk 6]
```diff
@@ -4,81 +4,284 @@
...
+          raise UnprocesableEntityError.new('Cannot add more layers to this visualization')
+        end
+        unless @map.admits_layer?(layer)
+          raise UnprocesableEntityError.new('Cannot add more layers of this type')
+        end
+
+        table_name = layer.options['table_name']
+        user_name = layer.options['user_name']
+        if user_name.present?
+          table_name = user_name + '.' + table_name
```
Comment (juanignaciosl) [inline app/controllers/carto/api/layers_controller.rb]: Shouldn't these names be quoted if needed?

File: app/controllers/carto/api/layers_controller.rb [hunk 7]
```diff
@@ -4,81 +4,284 @@
...
-        if @parent
-          parent = @parent
-          owner = @parent.user
+      def update
+        layers = @layers.map do |layer|
+          layer_params = params[:layers].present? ? params[:layers].find { |p| p['id'] == layer.id } : params
+
+          # don't allow to override table_name and user_name
+          new_layer_options = layer_params[:options]
+          if new_layer_options && new_layer_options.include?('table_name')
```
Comment (juanignaciosl) [inline app/controllers/carto/api/layers_controller.rb]: Minor: consider `new_layer_options.try(:include?, 'table_name')` (I've never been a big fan of Rails monkeypatched methods, but it looks like we're adopting them.

Comment (juanignaciosl) [inline app/controllers/carto/api/layers_controller.rb]: Minor 2: consider replacing the 2 checks with an iteration over `['user_name', 'table_name']`.

File: app/controllers/carto/api/layers_controller.rb [hunk 8]
```diff
@@ -4,81 +4,284 @@
...
+            new_layer_options['table_name'] = layer.options['table_name']
+          end
+          if new_layer_options && new_layer_options.include?('user_name')
+            new_layer_options['user_name'] = layer.options['user_name']
+          end
+
+          unless layer.update_attributes(layer_params.slice(:options, :kind, :infowindow, :tooltip, :order))
```
Comment (juanignaciosl) [inline app/controllers/carto/api/layers_controller.rb]: `:options, :kind, :infowindow, :tooltip, :order` is a duplication as well.

File: app/controllers/carto/api/layers_controller.rb [hunk 9]
```diff
@@ -4,81 +4,284 @@
...
+            raise UnprocesableEntityError.new(layer.errors.full_messages)
+          end
+
+          layer
+        end
+
+        if layers.count > 1
```
Comment (juanignaciosl) [inline app/controllers/carto/api/layers_controller.rb]: How come `update` can return either one or an array of Layers? :-\

File: app/controllers/carto/api/layers_controller.rb [hunk 10]
```diff
@@ -4,81 +4,284 @@
...
+          layers_json = layers.map { |l| Carto::Api::LayerPresenter.new(l, viewer_user: current_user).to_poro }
+          render_jsonp(layers: layers_json)
```
Comment (juanignaciosl) [inline app/controllers/carto/api/layers_controller.rb]: Inlining `layers_json` would make line sorter, in fact :-) .

File: app/controllers/carto/api/layers_controller.rb [hunk 11]
```diff
@@ -4,81 +4,284 @@
...
         else
-          parent = @owner_user
-          owner = @owner_user
+          render_jsonp Carto::Api::LayerPresenter.new(layers[0], viewer_user: current_user).to_poro
         end
+      rescue RuntimeError => e
+        CartoDB::Logger.error(
```
Comment (juanignaciosl) [inline app/controllers/carto/api/layers_controller.rb]: This fits in a line under 80 columns ;-)

File: app/controllers/carto/api/layers_controller.rb [hunk 12]
```diff
@@ -4,81 +4,284 @@
...
+        load_user_layers
+        raise LoadError.new('Layer not found') unless @layers.length == 1
+        @layer = @layers.first
+      end
 
-      def owner_from_params
-         @owner_user = current_user if (params[:user_id] && is_uuid?(params[:user_id]))
+      def load_user_layers
+        @layers = layers_ids.map { |id| @user.layers.find(id) }
+      rescue ActiveRecord::RecordNotFound
```
Comment (juanignaciosl) [inline app/controllers/carto/api/layers_controller.rb]: Maybe we could `recover_with` `ActiveRecord::RecordNotFound` as well instead of replacing it with `LoadError`, as nothing else is done.

File: app/controllers/carto/api/layers_controller.rb [hunk 13]
```diff
@@ -4,81 +4,284 @@
...
+        map_id = uuid_parameter(:map_id)
 
         # User must be owner or have permissions for the map's visualization
-        vis = Carto::Visualization.where({
-            map_id: params[:map_id]
-          }).first
-        raise RecordNotFound if vis.nil? || !vis.is_viewable_by_user?(current_user)
+        @map = Carto::Map.find(map_id)
+        vis = @map.visualization
+        raise LoadError.new('Map not found') unless vis && vis.is_viewable_by_user?(current_user)
```
Comment (juanignaciosl) [inline app/controllers/carto/api/layers_controller.rb]: Minor: consider `try`.

File: app/models/carto/map.rb [hunk 14]
```diff
@@ -148,8 +148,42 @@ def layer_selector
     options[:layer_selector]
   end
 
+  def admits_layer?(layer)
+    return admits_more_torque_layers? if layer.torque?
+    return admits_more_data_layers? if layer.data_layer?
+    return admits_more_base_layers?(layer) if layer.base_layer?
+  end
+
+  def can_add_layer(user)
```
Comment (juanignaciosl) [inline app/models/carto/map.rb]: Minor: `can_add_layer?`

File: app/models/carto/map.rb [hunk 15]
```diff
@@ -148,8 +148,42 @@ def layer_selector
...
+    if layer.uses_private_tables? && visualization.can_be_private?
+      visualization.privacy = Carto::Visualization::PRIVACY_PRIVATE
+      visualization.save
+    end
+  end
+
   private
 
+  def admits_more_data_layers?
+    data_layers.any? && visualization.canonical? ? false : true
```
Comment (juanignaciosl) [inline app/models/carto/map.rb]: This is equivalent to

`!(data_layers.any? && visualization.canonical?)`

File: app/models/carto/map.rb [hunk 16]
```diff
@@ -148,8 +148,42 @@ def layer_selector
...
+  end
+
+  def admits_more_torque_layers?
+    torque_layers.any? && visualization.canonical? ? false : true
```
Comment (juanignaciosl) [inline app/models/carto/map.rb]: Same: `!(torque_layers.any? && visualization.canonical?)`

File: spec/requests/api/json/layers_controller_spec.rb (hunk 1 above)
Comment (javitonino) [inline spec/requests/api/json/layers_controller_spec.rb]: I moved them to `spec/requests/api/json/layers_controller_shared_examples.rb`, as I used that to tests both controllers at the same time during development. The shared examples are ran from `layers_controller_spec` so it should be fine.

I also moved another two specs to shared_examples (map_layers.spec and user_layers.spec).

It may make sense now to merge the shared examples directly into the carto controller specs now, though, as they are only included from there in the final version.

File: app/controllers/carto/api/layers_controller.rb (hunk 6 above)
Comment (javitonino) [inline app/controllers/carto/api/layers_controller.rb]: They shouldn't. Quoting is SQL specific, so it should only be added when building an SQL query (and it is). Also, this is copied straight from the old controller.

File: app/controllers/carto/api/layers_controller.rb (hunk 3 above)
Comment (javitonino) [inline app/controllers/carto/api/layers_controller.rb]: When the layer is created and saved, it is not associated to a map, so we cannot add it directly to the Layer model validations.

We could try to add the validation to the map model, in a hook of the `layers` relationship, so it triggers when adding a new layer, but the problem there is that the layer is already created, so if the validation failed we should delete that layer. I think that is more complicated than the controller validation approach.

File: app/controllers/carto/api/layers_controller.rb (hunk 5 above)
Comment (javitonino) [inline app/controllers/carto/api/layers_controller.rb]: Thanks for bringing this up.

Yes and no. They need to be different in order to have different `before_filters`. We can use aliases instead of full methods in order to redirect to the common solution.

Another idea I had is using inheritance for this controllers, with a common LayersController, and two childs MapLayersController and UserLayersController:
 - Each child has different filters
 - There is a common implementation of index, show, create, update, destroy
 - We overload `create` adding map-specific validations, and then calling `super` for saving
 - `index` and `show` can probably share an implementation (the only difference is how they handle the owner user, and that can be handled in filters (`load_owner_user`))
 - `update` and `destroy` share the same implementation for sure

The same can be done with composition, but I feel inheritance is a bit better here). It adds some more changes, but I feel it might result in a cleaner controller. What do you think?

File: app/controllers/carto/api/layers_controller.rb (hunk 2 above)
Comment (javitonino) [inline app/controllers/carto/api/layers_controller.rb]: Good idea ✨

File: app/controllers/carto/api/layers_controller.rb (hunk 9 above)
Comment (javitonino) [inline app/controllers/carto/api/layers_controller.rb]: Updates can do 2 things for map layers:
 - If it receives a JSON object, it is a normal REST controller
 - If it receives `{ layers: [...] }`, it updates all layers in the array. This is used specially for reordering layers (but can be used for general bulk updates).

File: app/controllers/carto/api/layers_controller.rb (hunk 12 above)
Comment (javitonino) [inline app/controllers/carto/api/layers_controller.rb]: The messages for `ActiveRecord::RecordNotFound` change a bit depending on where it is raised. Do you think we could do without that, and just return a generic 404: Layer not found? It would be the code cleaner for sure.

File: app/models/carto/map.rb (hunk 15 above)
Comment (javitonino) [inline app/models/carto/map.rb]: Or even `data_layers.empty? || !visualization.canonical?`, but I feel that is more confusing. Changing to you sugestion.

**Comment by javitonino** [comment]
I like the suggestions, implementing them! I left a major discussion point, of which I think could be an interesting extra refactor step.

File: spec/requests/api/json/layers_controller_spec.rb (hunk 1 above)
Comment (juanignaciosl) [inline spec/requests/api/json/layers_controller_spec.rb]: Oh, ok. I took a quick look and new spec lines seemed much shorter than old ones. My bad!

File: app/controllers/carto/api/layers_controller.rb (hunk 5 above)
Comment (juanignaciosl) [inline app/controllers/carto/api/layers_controller.rb]: I also thought about inheritance. It fits here and it's clearer than aliases, but that's a matter of taste.

I'd refactor to inheritance, but maybe after merging this PR.

File: app/controllers/carto/api/layers_controller.rb (hunk 3 above)
Comment (juanignaciosl) [inline app/controllers/carto/api/layers_controller.rb]: Fair enough.

File: app/controllers/carto/api/layers_controller.rb (hunk 6 above)
Comment (juanignaciosl) [inline app/controllers/carto/api/layers_controller.rb]: 👍

File: app/controllers/carto/api/layers_controller.rb (hunk 9 above)
Comment (juanignaciosl) [inline app/controllers/carto/api/layers_controller.rb]: 👍

File: app/controllers/carto/api/layers_controller.rb (hunk 12 above)
Comment (juanignaciosl) [inline app/controllers/carto/api/layers_controller.rb]: Most of the messages that we raise `LoadError` with can be inferred from the request, so they're redundant. If the raised exception doesn't give more information I'd do it.

File: app/models/carto/map.rb (hunk 15 above)
Comment (juanignaciosl) [inline app/models/carto/map.rb]: Both are fine, maybe mine is clearer just because it's a more straightforward simplitication, yours look better on its own 👍

File: app/controllers/carto/api/layers_controller.rb (hunk 5 above)
Comment (javitonino) [inline app/controllers/carto/api/layers_controller.rb]: Ok, let's do it in another PR.

File: app/controllers/carto/api/layers_controller.rb (hunk 7 above)
Comment (javitonino) [inline app/controllers/carto/api/layers_controller.rb]: In fact, I'm going to rewrite this a bit, as right now, it removes `table_name` if not present in the request, which it shouldn't.

File: app/controllers/carto/api/layers_controller.rb [hunk 17]
```diff
@@ -4,81 +4,279 @@
...
+      # Takes a block, executed after saving
+      def save_layer(layer)
+        if layer.save
+          yield
+
+          render_jsonp Carto::Api::LayerPresenter.new(layer, viewer_user: current_user).to_poro
         else
-          parent = @owner_user
-          owner = @owner_user
+          CartoDB::Logger.info(message: 'Error creating layer', errors: layer.errors.full_messages)
```
Comment (juanignaciosl) [inline app/controllers/carto/api/layers_controller.rb]: Minor: how come is this an `info`?

File: app/controllers/carto/api/layers_controller.rb [hunk 18]
```diff
@@ -4,81 +4,279 @@
...
 
-        layer = parent.layers.where(id: params[:id]).first
-        raise RecordNotFound if layer.nil?
-        render_jsonp Carto::Api::LayerPresenter.new(layer, { viewer_user: current_user, user: owner}).to_json
+      def update
+        layers = @layers.map do |layer|
+          layer_params = params[:layers].present? ? params[:layers].find { |p| p['id'] == layer.id } : params
+
+          # don't allow to override table_name and user_name
+          new_layer_options = layer_params[:options]
```
Comment (juanignaciosl) [inline app/controllers/carto/api/layers_controller.rb]: Is this ok? It looks like `new_layer_options` is not used. The `delete` without a bang (!) doesn't seem right either.

File: app/controllers/carto/api/layers_controller.rb (hunk 17 above)
Comment (javitonino) [inline app/controllers/carto/api/layers_controller.rb]: Blind copy from old controller, probably better as a warning

File: app/controllers/carto/api/layers_controller.rb (hunk 18 above)
Comment (javitonino) [inline app/controllers/carto/api/layers_controller.rb]: `new_layer_options` is modified, and as it is a reference to `layer_params[:options]`, `layer_params` is modified. This is then used to update the layer.

`.delete!` doesn't exist. [`.delete`](https://docs.ruby-lang.org/en/2.0.0/Hash.html#method-i-delete) returns the value of the deleted key, and modified the existing hash.

There are two specs for this, so its behaviour should be ok.

Comment (juanignaciosl) [inline app/controllers/carto/api/layers_controller.rb]: Oh, sorry about the `delete` comment, I _always_ make that mistake. And it's true that content gets modified. It's a pity that it's not more explicit, but there's no point on making this more complicated 👍

**Comment by juanignaciosl** [comment]
Acceptance: IMHO this should be accepted through the smokes related to map edition, focusing on layers.

**Comment by javitonino** [comment]
Oh, I left a similar comment in the ticket (we should standardize where we write such comments). Copying:

> Acceptance: basically smoke tests for all things layer-related. Some must-test:
>- Test everything with Builder and Editor
>- Custom basemaps (create / delete)
>- Basemaps (test variants with layers on top and below)
>- Data and torque layers
>- Removing a dataset correctly identifies related maps
>- Cache/embed invalidation upon making changes (this is specially important in editor, which does not have the publish system)

**Comment by juanignaciosl** [comment]
@javitonino yes, as sometimes a PR has several issues, but we never do acceptance on several PRs, I'd add them at PRs. In addition, issues are the problem, PR the solution, so acceptance suggestions are closer to PRs, IMHO.

Anyway, sorry about not checking the issue!

**Comment by juanignaciosl** [comment]
@gfiorav spent some hours with many smokes: https://docs.google.com/spreadsheets/d/1_096XJQyYn83oj0xfpQxw88G82oQd52-Sbn1WpewbdY/edit#gid=1849408127 . We only found the `ssl_required` mistake ✨
